daemon/src/patch.rs

```diff
@@ -0,0 +1,169 @@
+    let mut patches = Vec::new();
+    let monorepo_path = state::monorepo(peer);
+    let monorepo = git2::Repository::open(monorepo_path)?;
+    let namespace = Namespace::from(project_urn.clone());
+    let default_branch_head_commit = {
+        let project = state::get_project(peer, project_urn.clone())
+            .await?
+            .ok_or_else(|| state::Error::ProjectNotFound(project_urn.clone()))?;
+        let maintainer = project
+            .delegations()
+            .iter()
+            .flat_map(|either| match either {
+                Either::Left(pk) => Either::Left(std::iter::once(PeerId::from(*pk))),
+                Either::Right(indirect) => {
+                    Either::Right(indirect.delegations().iter().map(|pk| PeerId::from(*pk)))
+                },
+            })
+            .next()
+            .expect("missing delegation");
+        // the `remote` for `get_branch` is set to the first maintainer, if the current `peer`
+        // is that maintainer, `get_branch` will catch that and search the local peers directories.
+        // The `branch` is set to `None` as `get_branch` will then fall back to the default branch.
+        let default_branch =
+            state::get_branch(peer, project_urn.clone(), Some(maintainer), None).await?;
+        monorepo
+            .find_reference(&default_branch.to_string())?
+            .peel_to_commit()?
+            .id()
+    };
+
+    for project_peer in state::list_project_peers(peer, project_urn.clone()).await? {
+        let remote = match project_peer {
+            project::Peer::Local { .. } => None,
+            project::Peer::Remote { peer_id, .. } => Some(peer_id),
+        };
+        let ref_pattern = Reference {
+            remote,
+            category: RefsCategory::Tags,
+            name: format!("{}*", TAG_PREFIX)
+                .parse::<RefspecPattern>()
+                .expect("invalid refspec pattern"),
+            namespace: Some(namespace.clone()),
+        };
+        let refs = ref_pattern.references(&monorepo)?;
+        for r in refs {
+            let r = r?;
+            match patch_from_ref(
+                &monorepo,
+                default_branch_head_commit,
+                project_peer.clone(),
+                &r,
+            ) {
+                Ok(patch) => patches.push(patch),
+                Err(err) => {
+                    log::warn!("failed to get patch from ref: {:?}", err);
+                },
+            }
+        }
+    }
+    Ok(patches)
+}
+
+#[derive(thiserror::Error, Debug)]
+enum PatchExtractError {
+    #[error("cannot peel reference to tag")]
+    PeelToTagFailed(#[source] git2::Error),
+    #[error("failed to determine merge base")]
+    FailedToDetermineMergeBase(#[source] git2::Error),
+    #[error("tag target object is not a commit")]
+    InvalidObjectType,
+    #[error("tag name is not valid UTF-8")]
+    InvalidName(#[source] std::str::Utf8Error),
+}
+
+fn patch_from_ref<'repo>(
```

Comment:
I'm wondering if some of this could expressed in terms of `Storage` rather than using the raw `monorepo`

Reply:
We need the monorepo to call `monorepo.merge_base()` and only for that. `Storage` does not provide that functionality yet. We could add it to `Storage` or use `Storage::as_raw()`.

In general, would you prefer, if we follow the pattern established in `project::checkout` where all functions take an `impl CanOpenStorage + 'static` parameter to access the monorepo?

Reply:
> In general, would you prefer, if we follow the pattern established in project::checkout where all functions take an impl CanOpenStorage + 'static parameter to access the monorepo?

Nope, that's not necessary because that's used for when interacting with remotes, which we're not doing here.

I think `as_raw` is available to any `librad::git` modules so that might be the best route to go down.

Reply:
`as_raw` is a crutch to avoid having to re-implement low-level git functionality for internal use. It is not intended to be part of the surface API.